### PR TITLE
8264386: LibraryLookup should be more friendly with implicit unloading

### PR DESCRIPTION
**Pull request by mcimadamore**
Quoting from the PR:

```
Dereferencing a segment from a LibraryLookup is possible, in the following form:

LibraryLookup libLookup = ...
MemorySegment segment = libLookup.lookup("foo").asSegmentRestricted(....)

The above code has a problem: there is no link between the produced segment and the original library lookup - meaning that, in principle, it's possible for the library to be unloaded when dereference occurs.
```

After having discussed many alternatives with Jorn, the best way to address this conundrum would be to add an overload of `LibraryLookup::lookup` which takes a lyout and returns a segment.

As part of this change I have also removed the LibraryLookup.Symbol API which seems unnecessary - now you can lookup an address or a segment. Of course, looking up a segment is creating a restricted segment (so it should be treated as unsafe).

I've added a test to make sure that the new API works as expected.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264386](https://bugs.openjdk.java.net/browse/JDK-8264386): LibraryLookup should be more friendly with implicit unloading

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - Committer) ⚠️ Review applies to 399d30b3eb74f15523f0bd700a52f74eb96b67d3
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/480/head:pull/480` \
`$ git checkout pull/480`

Update a local copy of the PR: \
`$ git checkout pull/480` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/480/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 480`

View PR using the GUI difftool: \
`$ git pr show -t 480`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/480.diff">https://git.openjdk.java.net/panama-foreign/pull/480.diff</a>

</details>

### CODE REVIEW

**Review by PaulSandoz — APPROVED** [review]
Summarizing so i have the correct understanding.

The core fix is to update `LibraryLookupImpl` to hold a memory segment that is sliced to the address of symbols, thereby keeping a ref-chain for returned segments **and** addresses of symbols.

Separately, the `Symbol` abstraction is not worth it, and we can lean on returning `MemorySegment` for a symbol covering some region of memory (like some constant).

**Review by JornVernee — APPROVED** [review]
LGTM, apart from the native access checks (see comments inline).

**Comment by mcimadamore** [comment]
> Summarizing so i have the correct understanding.
> 
> The core fix is to update `LibraryLookupImpl` to hold a memory segment that is sliced to the address of symbols, thereby keeping a ref-chain for returned segments **and** addresses of symbols.
> 
> Separately, the `Symbol` abstraction is not worth it, and we can lean on returning `MemorySegment` for a symbol covering some region of memory (like some constant).

Exactly. The only value added of Symbol was "the name getter" - but honestly that doesn't strike me as something super useful (given that you have to specify a name to the API in the first place).

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java [hunk 1]
```diff
@@ -103,29 +107,37 @@ static LibraryLookupImpl lookup(Supplier<NativeLibrary> librarySupplier, String
...
+            Reflection.ensureNativeAccess(Reflection.getCallerClass());
+            try {
+                Objects.requireNonNull(name);
+                Objects.requireNonNull(layout);
+                MemoryAddress addr = MemoryAddress.ofLong(library.lookup(name));
+                if (addr.toRawLongValue() % layout.byteAlignment() != 0) {
+                    throw new IllegalArgumentException("Bad layout alignment constraints: " + layout.byteAlignment());
+                }
+                return Optional.of(librarySegment.asSlice(addr)
+                        .asSlice(0L, layout.byteSize()));
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java:140]: my bad, when writing the code the IDE wasn't giving me that choice, so I assumed we forgot to add that combination, and moved on (with a view to file a bug). But now I see the overload is there...

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java [hunk 2]
```diff
@@ -103,29 +107,37 @@ static LibraryLookupImpl lookup(Supplier<NativeLibrary> librarySupplier, String
...
+                return Optional.of(librarySegment.asSlice(addr).address());
+            } catch (NoSuchMethodException ex) {
+                return Optional.empty();
+            }
+        }
+
+        @Override
+        @CallerSensitive
+        public Optional<MemorySegment> lookup(String name, MemoryLayout layout) {
+            Reflection.ensureNativeAccess(Reflection.getCallerClass());
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java:131]: @NativeAccess is for API only (e.g. think javadoc). So the annotations go in `LibraryLookup`

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java [hunk 3]
```diff
@@ -103,29 +107,37 @@ static LibraryLookupImpl lookup(Supplier<NativeLibrary> librarySupplier, String
...
         LibraryLookupImpl(NativeLibrary library, ResourceScope scope) {
             this.library = library;
-            this.scope = scope;
+            this.librarySegment = MemoryAddress.NULL.asSegmentRestricted(Long.MAX_VALUE, scope);
         }
 
         @Override
-        public Optional<Symbol> lookup(String name) {
+        public Optional<MemoryAddress> lookup(String name) {
             try {
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java:119]: in that case there's nothing that can go wrong, as we're not assigning fictional bounds to the result.

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/LibraryLookup.java [hunk 4]
```diff
@@ -66,39 +65,24 @@
...
-     * into a memory address (see {@link #name()} and {@link #address()}, respectively).
-     *
-     * @apiNote In the future, if the Java language permits, {@link Symbol}
-     * may become a {@code sealed} interface, which would prohibit subclassing except by
-     * explicitly permitted types.
-     *
-     * @implSpec
-     * Implementations of this interface are immutable, thread-safe and <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>.
+     * Looks up a symbol with given name in this library. The returned memory address maintains a strong reference to this lookup object.
+     * @param name the symbol name. This method can be useful to lookup function symbols in a foreign library.
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/LibraryLookup.java:69]: true

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java:131]: Ah ok, I forgot about the difference.

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**